heroku/cli
# fix: set default port for heroku:local to 5006

[work item](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001chAWvYAM/view)

This is a fix for #2515 that also does not re-introduce the defect found in #2457.

The approach to this fix is a bit unorthodox, so I am adding some notes here. It is not intended to be long-lived.

The [Node Foreman](https://www.npmjs.com/package/foreman) package, which is the basis for our `heroku:local` commands, sets the default port to 5000. However, Apple decided to use that port for AirPlay, which led to two PRs (#2261 and #2475) that set the default port for the local commands to 5001 using oclif defaults. However, this caused the current bug (#2515). Setting the oclif default overrode any .env files being read through Node Foreman.

Additionally complicating this matter, Node Foreman has not been maintained in several years. An [issue](https://github.com/strongloop/node-foreman/issues/173) was filed about this default port problem, but we have no confidence that it will be resolved.

We plan in the near future to remove our dependence on Node Foreman. In order to fix this problem in the short-term, however, we have copied over the `nf.js` file from Node Foreman to our own repo and updated the default port to 5006. We are using 5006 instead of 5001 because it is [not known to be used](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers) by other common software at this time.

## Testing
- Run `yarn install` and `yarn build`
- Navigate to the `/packages/cli` directory
- add a `.env` file that contains `PORT=3000`
- Run `./bin/dev local`
    - you should see the following:
```
[OKAY] Loaded ENV .env File as KEY=VALUE Format
11:20:53 AM web.1   |  it works (web)! port: 3000
11:20:53 AM worker.1 |  it works (worker)!
[DONE] Killing all processes with signal  SIGINT
11:20:53 AM web.1   Exited Successfully
11:20:53 AM worker.1 Exited Successfully
```
- Delete `PORT=3000` in the .env file inside `/packages/cli`
- Run `./bin/dev local`
   - you should see the following:
```
[OKAY] Loaded ENV .env File as KEY=VALUE Format
11:21:15 AM web.1   |  it works (web)! port: 5006
11:21:15 AM worker.1 |  it works (worker)!
[DONE] Killing all processes with signal  SIGINT
```
